text/0000-cfg-target.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: `cfg-target`
+- Start Date: 2020-09-27
+- RFC PR: [rust-lang/rfcs#2991](https://github.com/rust-lang/rfcs/pull/2991)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This proposes a new `cfg`: `target`, which matches the entire target triple
+string (e.g. `arm-unknown-linux-gnueabihf`). This also adds a `CARGO_CFG_TARGET`
+environment variable for parity with other `CARGO_CFG_*` variables.
+
+# Motivation
+[motivation]: #motivation
+
+To `#[cfg]` against a specific target, a `build.rs` script is required to emit a
+custom `cfg` based on the `TARGET` environment variable. Adding a build script
+increases compile time and makes a crate incompatible with certain build
+systems.
+
+Otherwise, all available components would need to be specified separately:
+`target_arch`, `target_vendor`, `target_os`, and `target_env`. This can be very
+cumbersome. Note that the target ABI cannot currently be `#[cfg]`-ed against, so
+a `build.rs` is still necessary to match all target components.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This would act like existing `target_*` configurations but match against all
+components (except `target_feature`).
+
+```rust
+#[cfg(target = "x86_64-apple-ios-macabi")]
+mod mac_catalyst;
+```
+
+This includes `#[cfg_attr(target = "...", attr)]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`target` is a key-value option set once with the target's Rust triple.
+
+Example values:
+
+- `"aarch64-apple-darwin"`
+- `"arm-unknown-linux-gnueabihf"`
+- `"x86_64-apple-ios-macabi"`
+- `"x86_64-pc-windows-gnu"`
+- `"x86_64-pc-windows-msvc"`
+- `"x86_64-unknown-linux-gnu"`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Configuring against specific targets can be overly strict and could make
+  certain `#[cfg]`s miss similar configurations with small changes.
+
+  For example: `aarch64-unknown-none` does not match
+  `aarch64-unknown-none-softfloat`, yet one would likely want to include ABI
+  variants. The same concern applies to the target vendor.
+
+  A potential solution would be to allow glob matching (e.g.
```

Comment:
How feasible would it be to allow specifying individual components of the target triple string by keyword inside a `target` cfg, something like:
```rust
#[cfg(target(arch="aarch64", os="none"))]
```

which would be equivalent to:
```rust
#[cfg(all(target_arch="aarch64", target_os="none"))]
```

?

I definitely agree with the assertion that "matching the entire target with the current set of `cfg` options is extremely verbose", but I'm not convinced that "allow matching the entire target string directly" is the right way to address that.